src/app/src/sensorUtils.js

```diff
@@ -17,46 +29,87 @@ export function makeRiverGaugeRequest(id, isApiRequest) {
+                );
+                return sensorValue !== apiVariableData.variable.noDataValue
+                    ? Object.assign(acc, { [variable]: sensorValue })
+                    : Object.assign(acc, { [variable]: 0 });
+            }
+            return acc;
         },
-    } = apiData;
-
-    const extractedVariableData = VARIABLES.reduce((acc, variable, idx) => {
-        const apiVariableData = data[idx];
-        if (apiVariableData) {
-            const sensorValue = Number(apiVariableData.values[0].value[0].value);
-            return sensorValue !== apiVariableData.variable.noDataValue
-                ? Object.assign(acc, {[variable]: sensorValue })
-                : Object.assign(acc, {[variable]: 0 });
-        }
-        return acc;
-    }, { id });
+        { id }
+    );
 
     return extractedVariableData;
 }
 
-export function parseRiverGaugeCsvData(id, csvString) {
-    const {
-        data,
-    } = csvString;
-
-    const parsedData = parseCsvString(data);
-    const dataRow = parsedData.data.slice(-1)[0];
+export function parseRiverGaugeCsvData(id, data) {
+    const dataRow = data.slice(-1)[0];
 
-    const extractedVariableData = VARIABLES.reduce((acc, variable) => {
-        const code = `p${VARIABLE_CODES[variable]}`;
-        return Object.assign(acc, {[variable]: dataRow[code] || 0 });
-    }, { id });
+    const extractedVariableData = VARIABLES.reduce(
+        (acc, variable) => {
+            const code = `p${VARIABLE_CODES[variable]}`;
+            return Object.assign(acc, { [variable]: dataRow[code] || 0 });
+        },
+        { id }
+    );
 
     return extractedVariableData;
 }
+
+export function getSensorByProp(prop, value) {
+    const sensor = find(sensors.features, f => {
+        return f.properties[prop] === value;
+    });
+
+    return sensor;
+}
+
+export function calculateOverallSensorRating(sensorRatings) {
+    const totalValue = values(sensorRatings).reduce((acc, r) => acc + r, 0);
+
+    if (totalValue >= 4) {
+        return RATING_GOOD;
+    } else if (totalValue >= 2) {
+        return RATING_FAIR;
+    } else {
+        return RATING_POOR;
+    }
+}
+
+export function transformSensorDataToRatings(sensorData) {
+    const sensor = getSensorByProp('Id', sensorData.id).properties;
+    const sensorRatings = VARIABLES.reduce((acc, variable) => {
+        const { lower, upper } = sensor.HealthyRanges[variable];
+        const variableValue = sensorData[variable];
+        const isVariableWithinHealthyRange =
+            lower <= variableValue && variableValue <= upper;
+
+        return Object.assign(acc, {
+            [variable]: isVariableWithinHealthyRange
+                ? VARIABLE_WITHIN_HEALTHY_RANGE
+                : VARIABLE_NOT_WITHIN_HEALTHY_RANGE,
+        });
+    }, {});
+
+    sensorRatings[OVERALL_RATING] = calculateOverallSensorRating(sensorRatings);
+
+    return { id: sensorData.id, sensorRatings: sensorRatings };
```

Comment:
Consider adjusting this to:

```
return {
   id: sensor.id,
   sensorRatings: {
       ...sensorRatings,
       [OVERALL_RATING]: calculateOverallSensorRating(sensorRatings),
   },
};
```

which would avoid the mutation in line 112.

Reply:
👍